### PR TITLE
Add information about ovn-octavia usage for Kuryr

### PR DESCRIPTION
**Pull request by luis5tb**
This PR adds information about OVN-octavia driver at Kuryr, and the benefits of it

### CODE REVIEW

**Comment by maxwelldb** [comment]
@luis5tb Thanks for this! To be clear, it applies to 4.4 and 4.5 w/no differences between the two?

**Comment by luis5tb** [comment]
> @luis5tb Thanks for this! To be clear, it applies to 4.4 and 4.5 w/no differences between the two?

Exactly, this is for 4.4 onwards

File: modules/installation-osp-about-kuryr.adoc [hunk 1]
```diff
@@ -40,7 +40,9 @@ OpenShift SDN over an OpenStack network.
 Conversely, Kuryr is not recommended in the following cases:
 
 * You use provider networks or tenant VLANs.
-* Your deployment uses many Services on a few hypervisors. Each OpenShift
-  Service creates an Octavia Amphora virtual machine in OpenStack that hosts a
-  required load balancer.
-* UDP Services are needed.
+* In {rh-openstack} versions earlier than 16, or if the {rh-openstack} Neutron
+  ML2/plugin is not OVN or ovn-octavia Octavia driver is not enabled, if:
```
Comment (MaysaMacedo) [inline modules/installation-osp-about-kuryr.adoc:44]: Perhaps might be good to replace the last `if` with `and your deployment uses:` and update the following line accordingly. Just to reinforce that the next points are the main concern when combined with version<16 / not ovn / not ovn-octavia.

File: modules/installation-osp-kuryr-octavia-configuration.adoc [hunk 2]
```diff
@@ -200,5 +200,40 @@ controller-0$ sudo docker restart octavia_worker
...
+| ovn     | Octavia OVN driver.                             |
++---------+-------------------------------------------------+
+----
+
+The OVN Octavia provider driver is an integration driver between load balancer
+provided by OVN and Octavia. It supports basic load balancer functionalities
+and is based on Openflow rules. This provider driver is automatically enabled
+in Octavia by Director on OVN Neutron ML2 enabled deployments. There are no
+additional installation or configuration steps required. The Amphora provider
+driver remains enabled and as the default provider driver.
```
Comment (MaysaMacedo) [inline modules/installation-osp-kuryr-octavia-configuration.adoc:230]: Isn't the ovn-octavia provider considered the default when enabled on osp16?

File: modules/installation-osp-kuryr-known-limitations.adoc [hunk 3]
```diff
@@ -7,20 +7,17 @@
...
-can cause you to run out of resources.
-
 [discrete]
 [id="openstack-version-limitations_{context}"]
 == {rh-openstack} version limitations
 
 Using {product-title} with Kuryr SDN has several limitations that depend on the {rh-openstack} version.
 
+* {rh-openstack} versions before 16 or without OVN octavia provider enabled uses
+the defatult Octavia load balancer driver (Amphora driver). This means an
```
Comment (MaysaMacedo) [inline modules/installation-osp-kuryr-known-limitations.adoc:17]: typo: default

**Review by MaysaMacedo — COMMENTED** [review]
Looks great. Just a few suggestions and question

File: modules/installation-osp-kuryr-octavia-configuration.adoc (hunk 2 above)
Comment (luis5tb) [reply modules/installation-osp-kuryr-octavia-configuration.adoc:230]: No, it will be the default for Kuryr. But in Octavia, the default will still be amphora driver. So, if you create a loadbalancer without specifying a provider, it will be an amphora load balancer

File: modules/installation-osp-kuryr-known-limitations.adoc [hunk 4]
```diff
@@ -7,20 +7,17 @@
...
-default Octavia load balancer driver (Amphora driver). Creating too many Services
-can cause you to run out of resources.
-
 [discrete]
 [id="openstack-version-limitations_{context}"]
 == {rh-openstack} version limitations
 
 Using {product-title} with Kuryr SDN has several limitations that depend on the {rh-openstack} version.
 
+* {rh-openstack} versions before 16 or without OVN octavia provider enabled uses
```
Comment (maxwelldb) [inline modules/installation-osp-kuryr-known-limitations.adoc:16]: This section seems largely to focus on RHOSP <16. Would it make sense for me to create a section for caveats that apply specifically to versions 13, 14, and 15 so that we don't need to keep stating "{rh-openstack} versions before 16..." here? @luis5tb

File: modules/installation-osp-default-kuryr-deployment.adoc [hunk 5]
```diff
@@ -26,7 +26,7 @@ Use the following quota to satisfy a default cluster's minimum requirements:
 |vCPUs                   | 28
 |Volume storage          | 275 GB
 |Instances               | 7
-|Security groups         | 250 - 1 needed per Service and per NetworkPolicy
+|Security groups         | 250 - 1 needed per Service (at {rh-openstack} version 16 with amphora driver instead of ovn octavia driver, this belongs to the service account instead of the user project) and per NetworkPolicy
```
Comment (maxwelldb) [inline modules/installation-osp-default-kuryr-deployment.adoc:29]: I'll move this into a note below the table so that it fits.

**Comment by maxwelldb** [comment]
Let me know what you think of edits on this module, @luis5tb: https://github.com/openshift/openshift-docs/pull/21487/commits/bc828b811bf89b0fb50830d9a8407eeeaa5b0c53

Rest is in-progress until I'm out of meetings.

File: modules/installation-osp-default-kuryr-deployment.adoc (hunk 5 above)
Comment (luis5tb) [reply modules/installation-osp-default-kuryr-deployment.adoc:29]: sounds good

File: modules/installation-osp-kuryr-known-limitations.adoc (hunk 4 above)
Comment (luis5tb) [reply modules/installation-osp-kuryr-known-limitations.adoc:16]: not really. this paragraph is about OSP13, 14, 15. But also about OSP16 with neutron ml2/ovs, or OSP16 with ml2/ovn but without ovn-octavia enabled

**Comment by luis5tb** [comment]
> @luis5tb to be clear:
> 
> (OSP <16 AND no ovn-octavia AND (UDP services OR many services)) OR many services?
> or
> (OSP <16 AND no ovn-octavia AND (UDP services OR many services))
> 

I would go for:
(OSP < 16 AND UDP services) OR ((OSP < 16 OR no ovn-octavia) AND many services)

> Do we want to attempt to define "many" in "many services?"

No, that depends on many things, size of the deployment, size of the amphora VM, overbooking ratio plus usage of those amphora, ...
> 
> And when we say "recommended," does this mean that such configurations are still supported?

Yes, we support OCP on OSP with Kuryr without ovn-octavia, and there are still advantages of using them (single overlay, no double encapsulation, posibility of exposing services (loadbalancer type) externally, ...) and of course one big con (need of a VM per svc)

**Comment by maxwelldb** [comment]
> > @luis5tb to be clear:
> > (OSP <16 AND no ovn-octavia AND (UDP services OR many services)) OR many services?
> > or
> > (OSP <16 AND no ovn-octavia AND (UDP services OR many services))
> 
> I would go for:
> (OSP < 16 AND UDP services) OR ((OSP < 16 OR no ovn-octavia) AND many services)
> 
> > Do we want to attempt to define "many" in "many services?"
> 
> No, that depends on many things, size of the deployment, size of the amphora VM, overbooking ratio plus usage of those amphora, ...
> 
> > And when we say "recommended," does this mean that such configurations are still supported?
> 
> Yes, we support OCP on OSP with Kuryr without ovn-octavia, and there are still advantages of using them (single overlay, no double encapsulation, posibility of exposing services (loadbalancer type) externally, ...) and of course one big con (need of a VM per svc)

Thanks!

Okay, so is this true?:

Kuryr is not recommended in deployments where:

* The {rh-openstack} version is less than 16.
* The deployment uses UDP services, or a large number of TCP services on few hypervisors.

or

* The `ovn-octavia` Octavia driver is disabled.
* The deployment uses a large number of TCP services on few hypervisors.

**Comment by maxwelldb** [comment]
@luis5tb See those last few commits for more edits. Any thoughts?

**Comment by luis5tb** [comment]
> > > @luis5tb to be clear:
> > > (OSP <16 AND no ovn-octavia AND (UDP services OR many services)) OR many services?
> > > or
> > > (OSP <16 AND no ovn-octavia AND (UDP services OR many services))
> > 
> > 
> > I would go for:
> > (OSP < 16 AND UDP services) OR ((OSP < 16 OR no ovn-octavia) AND many services)
> > > Do we want to attempt to define "many" in "many services?"
> > 
> > 
> > No, that depends on many things, size of the deployment, size of the amphora VM, overbooking ratio plus usage of those amphora, ...
> > > And when we say "recommended," does this mean that such configurations are still supported?
> > 
> > 
> > Yes, we support OCP on OSP with Kuryr without ovn-octavia, and there are still advantages of using them (single overlay, no double encapsulation, posibility of exposing services (loadbalancer type) externally, ...) and of course one big con (need of a VM per svc)
> 
> Thanks! @luis5tb
> 
> Okay, so is this true?:
> 
> Kuryr is not recommended in deployments where:
> 
> * The {rh-openstack} version is less than 16.
> * The deployment uses UDP services, or a large number of TCP services on few hypervisors.
> 
> or
> 
> * The `ovn-octavia` Octavia driver is disabled.
> * The deployment uses a large number of TCP services on few hypervisors.

Right

File: modules/installation-osp-kuryr-octavia-configuration.adoc [hunk 6]
```diff
@@ -200,5 +200,43 @@ controller-0$ sudo docker restart octavia_worker
 ====
 Depending on your OpenStack environment, Octavia might not support UDP
 listeners, which means there is no support for UDP Services if Kuryr SDN is
-used.
+used. {rh-openstack} version 16 and later supports Octavia.
```
Comment (luis5tb) [inline modules/installation-osp-kuryr-octavia-configuration.adoc:203]: support UDP, not Octavia

File: modules/installation-osp-kuryr-octavia-configuration.adoc [hunk 7]
```diff
@@ -200,5 +200,43 @@ controller-0$ sudo docker restart octavia_worker
...
+
+Beginning with {rh-openstack} version 16, the Octavia OVN provider driver (`ovn`) is supported on
+{product-title} on {rh-openstack} deployments.
+
+`ovn` is an integration driver for the load balancing
+that Octavia and OVN provide. It supports basic load balancing capabilities,
+and is based on OpenFlow rules. The driver is automatically enabled
+in Octavia by Director on deployments that use OVN Neutron ML2.
+
+If `ovn` is enabled, the Amphora provider
```
Comment (luis5tb) [inline modules/installation-osp-kuryr-octavia-configuration.adoc:235]: This is not true! Amphora provider still is the default one. The only different is that Kuryr will use the ovn one instead. But the default is amphora

File: modules/installation-osp-kuryr-octavia-configuration.adoc [hunk 8]
```diff
@@ -200,5 +200,43 @@ controller-0$ sudo docker restart octavia_worker
...
+driver remains available as a fallback driver.
+
+Kuryr offers better support for `ovn` than Amphora. If you choose `ovn`:
```
Comment (luis5tb) [inline modules/installation-osp-kuryr-octavia-configuration.adoc:238]: I think this is not the right way to put it. Kuryr supports both. But if OVN is available, kuryr will automatically use it, as OVN present some benefits (the above listed) for our use case (OCP on OSP)

**Comment by maxwelldb** [comment]
Couple of new commits for you, @luis5tb 
- https://github.com/openshift/openshift-docs/pull/21487/commits/5c74350fe805d6e246651d872808e9fb29a73aa7: Resolving your latest feedback
- https://github.com/openshift/openshift-docs/pull/21487/commits/0c51b4040cd414daa949a617c042df252b41a2dc: Starting to figure out the impact of [this comment](https://github.com/openshift/openshift-docs/pull/21487#commitcomment-38974115)

File: modules/installation-osp-kuryr-octavia-configuration.adoc [hunk 9]
```diff
@@ -198,7 +198,44 @@ controller-0$ sudo docker restart octavia_worker
 
 [NOTE]
 ====
-Depending on your OpenStack environment, Octavia might not support UDP
-listeners, which means there is no support for UDP Services if Kuryr SDN is
-used.
+Depending on your OpenStack environment, Octavia may not support UDP
+listeners. If you use Kuryr SDN on OpenStack version 15, UDP services are not supported.
```
Comment (luis5tb) [inline modules/installation-osp-kuryr-octavia-configuration.adoc:202]: 15 or earlier?

**Comment by luis5tb** [comment]
looks good, just one nit

File: modules/installation-osp-kuryr-octavia-configuration.adoc (hunk 9 above)
Comment (maxwelldb) [reply modules/installation-osp-kuryr-octavia-configuration.adoc:202]: Fixing.

**Comment by maxwelldb** [comment]
PTAL, @openshift/team-documentation

File: modules/installation-osp-default-kuryr-deployment.adoc [hunk 10]
```diff
@@ -40,22 +40,27 @@ A cluster might function with fewer than recommended resources, but its performa
...
 security group.
 
-* Each Service is mapped into an {rh-openstack} load balancer. Each load balancer
-has a security group with the user project; therefore, it must be taken into
-account when estimating the number of security groups required for the quota.
+* Each Service is mapped to an {rh-openstack} load balancer. Consider this requirement
+ when estimating the number of security groups required for the quota.
++
+If you are using
+{rh-openstack} versions 15 and earlier, or the `ovn-octavia driver`, each load balancer
```
Comment (bobfuru) [inline modules/installation-osp-default-kuryr-deployment.adoc:62]: s/versions 15 and earlier/version 15 or earlier/
Unless they could be running both 15 and earlier?

File: modules/installation-osp-kuryr-known-limitations.adoc [hunk 11]
```diff
@@ -7,20 +7,20 @@
...
-can cause you to run out of resources.
-
 [discrete]
 [id="openstack-version-limitations_{context}"]
 == {rh-openstack} version limitations
 
 Using {product-title} with Kuryr SDN has several limitations that depend on the {rh-openstack} version.
 
+* {rh-openstack} versions before 16 use
+the default Octavia load balancer driver (Amphora). This driver requires that an
```
Comment (bobfuru) [inline modules/installation-osp-kuryr-known-limitations.adoc:17]: s/an/one/ ?

File: modules/installation-osp-kuryr-octavia-configuration.adoc [hunk 12]
```diff
@@ -198,7 +198,44 @@ controller-0$ sudo docker restart octavia_worker
 
 [NOTE]
 ====
-Depending on your OpenStack environment, Octavia might not support UDP
-listeners, which means there is no support for UDP Services if Kuryr SDN is
-used.
+Depending on your OpenStack environment, Octavia may not support UDP
```
Comment (bobfuru) [inline modules/installation-osp-kuryr-octavia-configuration.adoc:201]: "may" - do you mean it is disallowed, or that it's possible? If the latter, then s/may/might.
**may** _v_
Use only to indicate permission. Otherwise, use “can” or “might.”

File: modules/installation-osp-kuryr-octavia-configuration.adoc [hunk 13]
```diff
@@ -198,7 +198,44 @@ controller-0$ sudo docker restart octavia_worker
...
+listeners. If you use Kuryr SDN on OpenStack version 15 or earlier, UDP services are not supported.
```
Comment (bobfuru) [inline modules/installation-osp-kuryr-octavia-configuration.adoc:202]: s/OpenStack/{rh-openstack}/

File: modules/installation-osp-kuryr-octavia-configuration.adoc [hunk 14]
```diff
@@ -198,7 +198,44 @@ controller-0$ sudo docker restart octavia_worker
...
+
+Beginning with {rh-openstack} version 16, the Octavia OVN provider driver (`ovn`) is supported on
+{product-title} on {rh-openstack} deployments.
+
+`ovn` is an integration driver for the load balancing
+that Octavia and OVN provide. It supports basic load balancing capabilities,
+and is based on OpenFlow rules. The driver is automatically enabled
+in Octavia by Director on deployments that use OVN Neutron ML2.
+
+The Amphora provider driver is the default driver. However, if `ovn` is enabled, Kuryr uses it.
```
Comment (bobfuru) [inline modules/installation-osp-kuryr-octavia-configuration.adoc:235]: s/However, if `ovn` is enabled, Kuryr uses it./If `ovn` is enabled, however, Kuryr uses it.

File: modules/installation-osp-kuryr-octavia-configuration.adoc [hunk 15]
```diff
@@ -198,7 +198,44 @@ controller-0$ sudo docker restart octavia_worker
...
+
+Using `ovn` with Kuryr instead of Amphora offers several benefits. If you choose `ovn`:
```
Comment (bobfuru) [inline modules/installation-osp-kuryr-octavia-configuration.adoc:237]: I'd suggest some rewording and front-loading, but it's subjective.

> Choosing `ovn` with Kuryr instead of Amphora offers several of the following benefits:
> * Decreased resource requirements. Kuryr does not require a load balancer VM for each service. 
> * Reduced network latency. 
> * Increased service creation speed using OpenFlow instead of a VM for each Service.
> * Full distribution of `ovn` work across all nodes.

**Review by bobfuru — COMMENTED** [review]
Just some small nits for consideration. Really good stuff overall. I didn't see a preview link, maybe because the PR was created during preview-bot downtime. Also, be sure to squash commits before merging. :)

File: modules/installation-osp-kuryr-octavia-configuration.adoc (hunk 12 above)
Comment (luis5tb) [reply modules/installation-osp-kuryr-octavia-configuration.adoc:201]: no no, it means it is not supported! It was not implemented at the time and Red Hat only supports it from OSP16

File: modules/installation-osp-kuryr-octavia-configuration.adoc (hunk 15 above)
Comment (luis5tb) [reply modules/installation-osp-kuryr-octavia-configuration.adoc:237]: I would reword it differently:
When Kuryr uses `ovn` instead of Amphora it offers the following benefits:

- Decreased resource requirements. Kuryr does not require a load balancer VM for each service.
- Reduced network latency
- Increased service creation speed using OpenFlow rules instead of a VM for each Service
- Distributed load balancing actions across all nodes, instead of centralized on the amphora VMs

**Comment by maxwelldb** [comment]
Thanks!

**Comment by maxwelldb** [comment]
@luis5tb If this all looks good, squash everything down into a single commit and force push. If you want any changes, let me know.

**Comment by luis5tb** [comment]
> @luis5tb If this all looks good, squash everything down into a single commit and force push. If you want any changes, let me know.
> 
> e: Fixed UL syntax in [e1ca589](https://github.com/openshift/openshift-docs/commit/e1ca5894c05199977fb0453ea0e4c26e80966b1f)

Looks good!
